File: Yoshino_ca_pv_conductivity.cpp
#include <cmath>
#include <stdexcept>
#include <iostream>

double Yoshino_ca_pv_conductivity(double temperature, double water_content, double pressure) {
    // Constants
    const double k_boltzmann = 1.380649e-23;  // Boltzmann constant in J/K
    const double J2eV = 1.0 / 1.60217733e-19; // 1 eV = 1.60217733e-19 J
    const double k = k_boltzmann * J2eV;      // Boltzmann constant in eV/K
    const double sigma_i = 631;               // Intrinsic conductivity in S/m
    const double Hi = 1.03;                   // Activation enthalpy for intrinsic conduction in eV
    const double sigma_h = 0;                 // Hydrogen-related conductivity in S/m
    const double Hh = 0;                      // Activation enthalpy for hydrogen in eV
    const double sigma_p = 0;                 // Proton-related conductivity in S/m
    const double Hp = 0;                      // Activation enthalpy for proton conduction in eV
    const double alpha = 0;                   // Constant for proton conduction

    // Input validation
    if (temperature <= 0) {
        throw std::invalid_argument("Temperature must be greater than zero Kelvin.");
    }

    if (pressure < 0) {
        throw std::invalid_argument("Pressure must be non-negative.");
    }

    if (water_content < 0 || water_content > 1) {
        throw std::invalid_argument("Water content must be in the range of 0 to 1.");
    }

    // Convert pressure from Pa to J/cc (1 Pa = 1 J/m³ = 1e-6 J/cc)
    pressure *= 1e3;

    // Sum the conductivities
    double sigma = sigma_i * std::exp(-Hi / (k * temperature)) +
                   sigma_h * std::exp(-Hh / (k * temperature)) +
                   sigma_p * water_content * std::exp(-(Hp - alpha * std::pow(water_content, 1.0 / 3.0)) / (k * temperature));

    return sigma;
}

// Example usage
int main() {
    double temperature = 1000.0;  // Example temperature in Kelvin
    double water_content = 0.1;   // Example water content
    double pressure = 1e5;        // Example pressure in Pascal

    try {
        double conductivity = Yoshino_ca_pv_conductivity(temperature, water_content, pressure);
        std::cout << "Conductivity: " << conductivity << " S/m" << std::endl;
    } catch (const std::invalid_argument& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }

    return 0;
}

File: matlab-version/Yoshino_ca_pv_conductivity.m
%Xu Y, McCammon C, Poe B T. The effect of alumina on the electrical conductivity of silicate perovskite[J]. 
%Science, 1998, 282(5390): 922-924.

function sigma = Yoshino_ca_pv_conductivity(temperature, water_content,pressure)
    % Constants
    % This function calculates the electrical conductivity based on the
    % provided temperature (T), water content (Cw), and material constants.
    % Constants
    k = 1.380649*10^(-23);
    J2eV = 1.0/(1.60217733*10^(-19)); % 1eV = 1.60217733*10^(-19) J
    k = k*J2eV;
    
    sigma_i = 631;
    Hi = 1.03;
    sigma_h = 0;
    Hh = 0;
    sigma_p = 0;
    Hp = 0;
    alpha = 0;
    T =  temperature;
    Cw = water_content;     
    % Check if temperature is greater than zero
    if any(temperature <= 0)
        error('Temperature must be greater than zero Kelvin.');
    end

    % Ensure pressure is numeric
    if iscell(pressure)
        pressure = cell2mat(pressure); % Convert cell array to numeric if necessary
    end

    % Check if pressure is non-negative
    if any(pressure < 0)
        error('Pressure must be non-negative.');
    end

    % Convert pressure from Pa to J/cc
    pressure = 1e3 * pressure;  % 1 Pa = 1 J/m³ = 1e-6 J/cc

    % Ensure water_content is numeric
    if iscell(water_content)
        water_content = cell2mat(water_content); % Convert cell array to numeric if necessary
    end

    % Check if water_content is within the range of 0 to 1
    if any(water_content < 0) || any(water_content > 1)
        error('Water content must be in the range of 0 to 1.');
    end

    % Sum the conductivities
    sigma = sigma_i*exp(-Hi/(k*T))+...
        sigma_h*exp(-Hh/(k*T))+...
        sigma_p*Cw*exp(-(Hp-alpha*Cw^(1.0/3.0))/(k*T));
end
